src/Service.Tests/Unittests/MultipleCreateUnitTests/MultipleCreateOrderHelperUnitTests.cs

```diff
@@ -0,0 +1,399 @@
+        /// relationship fields are non-autogenerated.
+        /// </summary>
+        [TestMethod]
+        public void ValidateIndeterministicReferencingEntityForAutogenAndNonAutogenRelationshipColumns()
+        {
+            // Test 1
+            IMiddlewareContext context = SetupMiddlewareContext();
+            string sourceEntityName = "User_AutogenToNonAutogenRelationshipColumn";
+            string targetEntityName = "UserProfile";
+
+            // Setup column input in source entity.
+            Dictionary<string, IValueNode> columnDataInSourceBody = new()
+            {
+                { "email", new StringValueNode("dab@microsoft.com") }
+            };
+
+            // Setup column input in target entity.
+            ObjectValueNode targetNodeValue = new();
+            List<ObjectFieldNode> fields = new()
+            {
+                new ObjectFieldNode("profilepictureurl", "dab/profilepicture"),
+                new ObjectFieldNode("username", "DAB")
+            };
+            targetNodeValue = targetNodeValue.WithFields(fields);
+
+            // Since the source entity contains an autogenerated relationship field (userid) and the input for target entity
+            // contains the relationship field 'username' in it, assert that we get the expected exception.
+            DataApiBuilderException ex = Assert.ThrowsException<DataApiBuilderException>(() => MultipleCreateOrderHelper.GetReferencingEntityName(
+                relationshipName: "UserProfile_AutogenToNonAutogenRelationshipColumn",
+                context: context,
+                sourceEntityName: sourceEntityName,
+                targetEntityName: targetEntityName,
+                metadataProvider: _sqlMetadataProvider,
+                columnDataInSourceBody: columnDataInSourceBody,
+                targetNodeValue: targetNodeValue,
+                nestingLevel: 1));
+        }
+
+        /// <summary>
+        /// Test to validate that when a relationship field in one of the entity is autogenerated, then if the value
+        /// of all other non-autogenerated fields is also specified in the input for the same entity, we successfully determine that
+        /// the other entity will act as the referencing entity. This is because the first entity, on its own, will contain values for
+        /// all the columns required to do insertion and act as a referenced entity.
+        ///
+        /// Here, the relationship between entities: 'User_AutogenToNonAutogenRelationshipColumn' and
+        /// 'UserProfile_NonAutogenToAutogenRelationshipColumn' is defined as:
+        /// User_AutogenToNonAutogenRelationshipColumn(userid, username) -> UserProfile_NonAutogenToAutogenRelationshipColumn(userid, username)
+        /// where both the relationships fields User_AutogenToNonAutogenRelationshipColumn.userid is an autogenerated field while all other
+        /// relationship fields are non-autogenerated.
```

Comment:
> where both the relationships fields User_AutogenToNonAutogenRelationshipColumn.userid is an autogenerated field
both? only one field is named.

I'm assuming here that you mean `User_...`.userid is the only autogenerated field. Because the autogenerated field is on `User_..` that means that all other relationship fields on `User_..` which is `username` MUST have a value supplied on the source entity, `User_..`. If the value of `username` were supplied for the target entity in this case, the insertion order is indeterministic?

Reply:
Correct. I'm correcting the comment.